File: bot-api/python/src/robocode_tank_royale/bot_api/mapper/event_mapper.py
import json
from typing import Any, Dict, Union

from ..base_bot_abc import BaseBotABC
from ..bot_exception import BotException
from ..events import (
    BotDeathEvent,
    BulletFiredEvent,
    BulletHitBotEvent,
    BulletHitBulletEvent,
    BulletHitWallEvent,
    DeathEvent,
    HitBotEvent,
    HitByBulletEvent,
    HitWallEvent,
    ScannedBotEvent,
    SkippedTurnEvent,
    TeamMessageEvent,
    TickEvent,
    WonRoundEvent, BotEvent
)
from robocode_tank_royale.schema import (
    BotDeathEvent as SchemaBotDeathEvent,
    BotHitBotEvent as SchemaBotHitBotEvent,
    BotHitWallEvent as SchemaBotHitWallEvent,
    BulletFiredEvent as SchemaBulletFiredEvent,
    BulletHitBotEvent as SchemaBulletHitBotEvent,
    BulletHitBulletEvent as SchemaBulletHitBulletEvent,
    BulletHitWallEvent as SchemaBulletHitWallEvent,
    Event as SchemaEvent,
    ScannedBotEvent as SchemaScannedBotEvent,
    SkippedTurnEvent as SchemaSkippedTurnEvent,
    TeamMessageEvent as SchemaTeamMessageEvent,
    TickEventForBot as SchemaTickEvent,
    WonRoundEvent as SchemaWonRoundEvent
)
from .bot_state_mapper import BotStateMapper
from .bullet_state_mapper import BulletStateMapper


class EventMapper:
    """Utility class for mapping events."""

    @staticmethod
    def map_tick_event(event: SchemaTickEvent, base_bot: BaseBotABC) -> TickEvent:
        """Map a schema TickEventForBot to a bot-api TickEvent."""
        mapped_events = None
        assert event.events is not None
        non_null_events: list[SchemaEvent] = []
        for e in event.events:
            assert e is not None
            non_null_events.append(e)
        mapped_events = EventMapper._map_events(non_null_events, base_bot)
        assert mapped_events is not None
        return TickEvent(
            event.turn_number,
            event.round_number,
            BotStateMapper.map(event.bot_state),
            BulletStateMapper.map(event.bullet_states),
            mapped_events,
        )

    @staticmethod
    def _map_events(events: list[SchemaEvent], base_bot: BaseBotABC) -> list[BotEvent] | None:
        """Map a collection of schema events to bot-api events."""
        game_bot_events: list[BotEvent] = list()
        for event in events:
            game_bot_events.append(EventMapper.map_event(event, base_bot))
        return game_bot_events

    @staticmethod
    def map_event(event: SchemaEvent, base_bot: BaseBotABC) -> BotEvent:
        """Map a schema event to a bot-api event."""
        event_type = type(event).__name__

        if isinstance(event, SchemaBotDeathEvent):
            return EventMapper._map_bot_death_event(event, base_bot.get_my_id())
        if isinstance(event, SchemaBotHitBotEvent):
            return EventMapper._map_hit_bot_event(event)
        if isinstance(event, SchemaBotHitWallEvent):
            return EventMapper._map_hit_wall_event(event)
        if isinstance(event, SchemaBulletFiredEvent):
            return EventMapper._map_bullet_fired_event(event)
        if isinstance(event, SchemaBulletHitBotEvent):
            return EventMapper._map_bullet_hit_bot_event(event, base_bot.get_my_id())
        if isinstance(event, SchemaBulletHitBulletEvent):
            return EventMapper._map_bullet_hit_bullet_event(event)
        if isinstance(event, SchemaBulletHitWallEvent):
            return EventMapper._map_bullet_hit_wall_event(event)
        if isinstance(event, SchemaScannedBotEvent):
            return EventMapper._map_scanned_bot_event(event)
        if isinstance(event, SchemaSkippedTurnEvent):
            return EventMapper.map_skipped_turn_event(event)
        if isinstance(event, SchemaWonRoundEvent):
            return EventMapper._map_won_round_event(event)
        if isinstance(event, SchemaTeamMessageEvent):
            return EventMapper._map_team_message_event(event)

        raise BotException(f"No mapping exists for event type: {event_type}")

    @staticmethod
    def _map_bot_death_event(source: SchemaBotDeathEvent, my_bot_id: int) -> Union[DeathEvent, BotDeathEvent]:
        """Map a schema BotDeathEvent to a bot-api event."""
        if source.victim_id == my_bot_id:
            return DeathEvent(source.turn_number)
        return BotDeathEvent(source.turn_number, source.victim_id)

    @staticmethod
    def _map_hit_bot_event(source: SchemaBotHitBotEvent) -> HitBotEvent:
        """Map a schema BotHitBotEvent to a bot-api HitBotEvent."""
        return HitBotEvent(
            source.turn_number,
            source.victim_id,
            source.energy,
            source.x,
            source.y,
            source.rammed
        )

    @staticmethod
    def _map_hit_wall_event(source: SchemaBotHitWallEvent) -> HitWallEvent:
        """Map a schema BotHitWallEvent to a bot-api HitWallEvent."""
        return HitWallEvent(source.turn_number)

    @staticmethod
    def _map_bullet_fired_event(source: SchemaBulletFiredEvent) -> BulletFiredEvent:
        """Map a schema BulletFiredEvent to a bot-api BulletFiredEvent."""
        return BulletFiredEvent(
            source.turn_number,
            BulletStateMapper.map(source.bullet)
        )

    @staticmethod
    def _map_bullet_hit_bot_event(source: SchemaBulletHitBotEvent, my_bot_id: int) -> Union[
        HitByBulletEvent, BulletHitBotEvent]:
        """Map a schema BulletHitBotEvent to a bot-api event."""
        bullet = BulletStateMapper.map(source.bullet)
        if source.victim_id == my_bot_id:
            return HitByBulletEvent(
                source.turn_number,
                bullet,
                source.damage,
                source.energy
            )
        return BulletHitBotEvent(
            source.turn_number,
            source.victim_id,
            bullet,
            source.damage,
            source.energy
        )

    @staticmethod
    def _map_bullet_hit_bullet_event(source: SchemaBulletHitBulletEvent) -> BulletHitBulletEvent:
        """Map a schema BulletHitBulletEvent to a bot-api BulletHitBulletEvent."""
        return BulletHitBulletEvent(
            source.turn_number,
            BulletStateMapper.map(source.bullet),
            BulletStateMapper.map(source.hit_bullet)
        )

    @staticmethod
    def _map_bullet_hit_wall_event(source: SchemaBulletHitWallEvent) -> BulletHitWallEvent:
        """Map a schema BulletHitWallEvent to a bot-api BulletHitWallEvent."""
        return BulletHitWallEvent(
            source.turn_number,
            BulletStateMapper.map(source.bullet)
        )

    @staticmethod
    def _map_scanned_bot_event(source: SchemaScannedBotEvent) -> ScannedBotEvent:
        """Map a schema ScannedBotEvent to a bot-api ScannedBotEvent."""
        return ScannedBotEvent(
            source.turn_number,
            source.scanned_by_bot_id,
            source.scanned_bot_id,
            source.energy,
            source.x,
            source.y,
            source.direction,
            source.speed
        )

    @staticmethod
    def map_skipped_turn_event(source: SchemaSkippedTurnEvent) -> SkippedTurnEvent:
        """Map a schema SkippedTurnEvent to a bot-api SkippedTurnEvent."""
        return SkippedTurnEvent(source.turn_number)

    @staticmethod
    def _map_won_round_event(source: SchemaWonRoundEvent) -> WonRoundEvent:
        """Map a schema WonRoundEvent to a bot-api WonRoundEvent."""
        return WonRoundEvent(source.turn_number)

    @staticmethod
    def _map_team_message_event(source: SchemaTeamMessageEvent) -> TeamMessageEvent:
        """Map a schema TeamMessageEvent to a bot-api TeamMessageEvent."""
        message = source.message
        if message is None:  # type: ignore
            raise BotException("message in TeamMessageEvent is None")

        try:
            message_object: Dict[str, Any] = json.loads(message)
            return TeamMessageEvent(source.turn_number, message_object, source.sender_id)
        except json.JSONDecodeError as e:
            raise BotException(f"Could not parse team message: {str(e)}")

File: bot-api/typescript/src/mapper/EventMapper.ts
import {
  TickEventForBot,
  TickEvent as SchemaTickEvent,
  BotDeathEvent as SchemaBotDeathEvent,
  BotHitBotEvent as SchemaBotHitBotEvent,
  BotHitWallEvent as SchemaBotHitWallEvent,
  BulletFiredEvent as SchemaBulletFiredEvent,
  BulletHitBotEvent as SchemaBulletHitBotEvent,
  BulletHitBulletEvent as SchemaBulletHitBulletEvent,
  BulletHitWallEvent as SchemaBulletHitWallEvent,
  ScannedBotEvent as SchemaScannedBotEvent,
  WonRoundEvent as SchemaWonRoundEvent,
  TeamMessageEvent as SchemaTeamMessageEvent,
  SkippedTurnEvent as SchemaSkippedTurnEvent,
} from "../protocol/schema.js";
import { MessageType } from "../protocol/MessageType.js";
import { BotEvent } from "../events/BotEvent.js";
import { TickEvent } from "../events/TickEvent.js";
import { DeathEvent } from "../events/DeathEvent.js";
import { BotDeathEvent } from "../events/BotDeathEvent.js";
import { HitBotEvent } from "../events/HitBotEvent.js";
import { HitWallEvent } from "../events/HitWallEvent.js";
import { BulletFiredEvent } from "../events/BulletFiredEvent.js";
import { HitByBulletEvent } from "../events/HitByBulletEvent.js";
import { BulletHitBotEvent } from "../events/BulletHitBotEvent.js";
import { BulletHitBulletEvent } from "../events/BulletHitBulletEvent.js";
import { BulletHitWallEvent } from "../events/BulletHitWallEvent.js";
import { ScannedBotEvent } from "../events/ScannedBotEvent.js";
import { WonRoundEvent } from "../events/WonRoundEvent.js";
import { TeamMessageEvent } from "../events/TeamMessageEvent.js";
import { SkippedTurnEvent } from "../events/SkippedTurnEvent.js";
import { BotStateMapper } from "./BotStateMapper.js";
import { BulletStateMapper } from "./BulletStateMapper.js";

/** Maps a schema TickEventForBot to a API TickEvent with all sub-events mapped. */
export class EventMapper {
  static map(tick: TickEventForBot, myBotId: number): TickEvent {
    const events = tick.events.map((e) => EventMapper.mapEvent(e, myBotId)).filter((e): e is BotEvent => e !== null);
    return new TickEvent(
      tick.turnNumber,
      tick.roundNumber,
      BotStateMapper.map(tick.botState),
      BulletStateMapper.mapCollection(tick.bulletStates),
      events,
    );
  }

  private static mapEvent(e: SchemaTickEvent, myBotId: number): BotEvent | null {
    switch (e.type) {
      case MessageType.BotDeathEvent: {
        const ev = e as SchemaBotDeathEvent;
        if (ev.victimId === myBotId) {
          return new DeathEvent(ev.turnNumber);
        }
        return new BotDeathEvent(ev.turnNumber, ev.victimId);
      }
      case MessageType.BotHitBotEvent: {
        const ev = e as SchemaBotHitBotEvent;
        return new HitBotEvent(ev.turnNumber, ev.victimId, ev.energy, ev.x, ev.y, ev.rammed);
      }
      case MessageType.BotHitWallEvent: {
        const ev = e as SchemaBotHitWallEvent;
        return new HitWallEvent(ev.turnNumber);
      }
      case MessageType.BulletFiredEvent: {
        const ev = e as SchemaBulletFiredEvent;
        return new BulletFiredEvent(ev.turnNumber, BulletStateMapper.map(ev.bullet));
      }
      case MessageType.BulletHitBotEvent: {
        const ev = e as SchemaBulletHitBotEvent;
        if (ev.victimId === myBotId) {
          return new HitByBulletEvent(ev.turnNumber, BulletStateMapper.map(ev.bullet), ev.damage, ev.energy);
        }
        return new BulletHitBotEvent(ev.turnNumber, ev.victimId, BulletStateMapper.map(ev.bullet), ev.damage, ev.energy);
      }
      case MessageType.BulletHitBulletEvent: {
        const ev = e as SchemaBulletHitBulletEvent;
        return new BulletHitBulletEvent(ev.turnNumber, BulletStateMapper.map(ev.bullet), BulletStateMapper.map(ev.hitBullet));
      }
      case MessageType.BulletHitWallEvent: {
        const ev = e as SchemaBulletHitWallEvent;
        return new BulletHitWallEvent(ev.turnNumber, BulletStateMapper.map(ev.bullet));
      }
      case MessageType.ScannedBotEvent: {
        const ev = e as SchemaScannedBotEvent;
        return new ScannedBotEvent(ev.turnNumber, ev.scannedByBotId, ev.scannedBotId, ev.energy, ev.x, ev.y, ev.direction, ev.speed);
      }
      case MessageType.WonRoundEvent: {
        const ev = e as SchemaWonRoundEvent;
        return new WonRoundEvent(ev.turnNumber);
      }
      case MessageType.TeamMessageEvent: {
        const ev = e as SchemaTeamMessageEvent;
        return new TeamMessageEvent(ev.turnNumber, ev.message, ev.senderId);
      }
      case MessageType.SkippedTurnEvent: {
        const ev = e as SchemaSkippedTurnEvent;
        return new SkippedTurnEvent(ev.turnNumber);
      }
      default:
        return null;
    }
  }
}
